rats-apps/test/python/rats_test/apps/test_annotated_container.py

```diff
@@ -0,0 +1,288 @@
+from collections.abc import Callable
+from typing import Any, NamedTuple, ParamSpec
+
+import pytest
+
+from rats import apps
+
+
+class ExampleStorage:
+
+    called: int
+    data: list[str]
+
+    def __init__(self) -> None:
+        self.called = 0
+        self.data = []
+
+    def save(self, data: str) -> None:
+        self.called += 1
+        self.data.append(data)
+
+
+class ExampleConfig(NamedTuple):
+    account: str
+    table: str
+    max: int
+
+
+class ExampleExecutable:
+
+    called: int
+
+    def __init__(self) -> None:
+        self.called = 0
+
+    def execute(self) -> None:
+        self.called += 1
+
+
+P_ProviderParams = ParamSpec("P_ProviderParams")
+
+
+def custom(callback: Callable[[P_ProviderParams], apps.ServiceId]) -> Callable[[P_ProviderParams], apps.ServiceId]:
+    def run(*args, **kwargs) -> apps.ServiceId:
+        return callback(*args, **kwargs)
+
+    return run
+
+
+class ExampleGroups:
+    STORAGE_REPLICAS = apps.ServiceId[ExampleStorage]("/group/storage-replicas")
+    EXECUTABLES = apps.ServiceId[ExampleExecutable]("/group/exe")
+    MISSING = apps.ServiceId[ExampleStorage]("/group/missing")
+
+
+class ExampleIds:
+    STORAGE = apps.ServiceId[ExampleStorage]("/service/storage")
+    EXECUTABLE = apps.ServiceId[ExampleExecutable]("/service/exe")
+    MISSING = apps.ServiceId[Any]("/service/missing")
+    DUPLICATE = apps.ServiceId[Any]("/service/duplicate")
+    GROUPS = ExampleGroups
+
+    make_service_id = custom(lambda x: apps.ServiceId[Any](f"/generated/{x}"))
+    make_config_id = custom(lambda x: apps.ConfigId[Any](f"/generated/config-{x}"))
+
+
+class ExampleContainer(apps.AnnotatedContainer):
+
+    @apps.service(ExampleIds.STORAGE)
+    def storage(self) -> ExampleStorage:
+        return ExampleStorage()
+
+    @apps.service(ExampleIds.DUPLICATE)
+    def duplicate_1(self) -> ExampleStorage:
+        return ExampleStorage()
+
+    @apps.service(ExampleIds.DUPLICATE)
+    def duplicate_2(self) -> ExampleStorage:
+        return ExampleStorage()
+
+    @apps.group(ExampleIds.GROUPS.STORAGE_REPLICAS)
+    def replica_1(self) -> ExampleStorage:
+        return ExampleStorage()
+
+    @apps.group(ExampleIds.GROUPS.STORAGE_REPLICAS)
+    def replica_2(self) -> ExampleStorage:
+        return ExampleStorage()
+
+    @apps.container()
```

Comment:
does this register the instance of ExamplePlugin such that when one searches for service A in this ExampleContainer instance, it would be searched for in the ExamplePlugin?

Reply:
yes, you can think of all the child containers as being merged together into one bucket of ids